feat(#58): Order 데이터베이스 스키마 및 JPA 리포지토리 구현

joel-you opened the pull request:
## 📋 PR 요약

Order Service의 데이터베이스 스키마 설계 및 JPA Repository 계층을 구현했습니다. Hexagonal Architecture 기반으로 도메인 모델과 영속성 계층을 분리하고, MapStruct를 활용한 엔티티-도메인 변환을 적용했습니다.

### 🔗 관련 이슈

- Closes #58

### 📝 변경 사항

- [x] 새로운 기능 추가
- [ ] 버그 수정
- [ ] 리팩토링
- [ ] 성능 개선
- [ ] 문서 업데이트
- [x] 테스트 코드 추가/수정
- [ ] 기타:

### 🛠️ 기술적 변경 사항

**수정된 파일**:
- `OrderRepositoryImpl.java` - Repository 구현 완료 (15개 메서드)
- `OrderJpaEntity.java` - 컬럼명 업데이트 (orderUuid→orderId, userId→customerId)
- `OrderDomainMapper.java` - MapStruct 매퍼 인터페이스 정의

**새로 추가된 파일**:
- `infra/db/migration/V1__Create_order_tables.sql` - Flyway 마이그레이션 스크립트
- `OrderRepositoryImplTest.java` - Repository 단위 테스트 (19개 테스트 케이스)
- 테스트 파일 수정: `OrderJpaEntityTest.java`, `OrderItemJpaEntityTest.java`

**제거된 파일**:
- `order-orchestrator/src/main/resources/db/migration/V1__Create_order_tables.sql` (infra로 이동)

**주요 로직 변경**:
1. **Repository 구현 완료**
    - 저장/조회/삭제 메서드 15개 구현
    - MapStruct 기반 엔티티 ↔ 도메인 자동 변환
    - N+1 문제 방지를 위한 JOIN FETCH 쿼리

2. **데이터베이스 스키마**
    - `orders`, `order_items` 테이블 생성
    - UUID 기반 주문 식별자 적용
    - 성능 최적화 인덱스 추가 (customer_id, status, order_date 등)

3. **테스트 전략**
    - Mock 기반 단위 테스트
    - 성공/실패 시나리오 분리
    - 도메인 변환 검증

### 🧪 테스트

- [x] 단위 테스트 통과 (103개 테스트, 100% 통과)
- [ ] 통합 테스트 통과
- [ ] E2E 테스트 통과 (해당하는 경우)
- [x] 수동 테스트 완료

**테스트 방법:**

1. `./gradlew :order-orchestrator:test` 실행
2. 전체 103개 테스트 통과 확인
3. Coverage Report 확인:
    - Line Coverage: 100.00%
    - Branch Coverage: 66.67%

**테스트 범위:**
- OrderRepositoryImpl: 19개 테스트
- OrderJpaEntity: 2개 테스트
- OrderItemJpaEntity: 2개 테스트
- Order 도메인: 30개 테스트
- OrderItem 도메인: 25개 테스트
- OrderStatus: 15개 테스트
- OrderDomainMapper: 2개 테스트

### 📊 성능 영향

- [x] 성능 향상 있음
- [ ] 성능에 영향 없음
- [ ] 성능 저하 가능성 있음

**성능 최적화 사항:**
- LEFT JOIN FETCH를 통한 N+1 문제 방지
- 복합 인덱스 적용 (customer_id + status)
- 날짜 범위 조회 인덱스 최적화

### 🔒 보안 체크리스트

- [x] 새로운 의존성 추가 시 보안 검토 완료 (MapStruct만 추가)
- [x] 사용자 입력 검증 로직 확인 (도메인 레벨 검증 존재)
- [x] 인증/권한 로직 검토 완료 (Repository 계층에는 미포함)
- [x] 민감한 정보 노출 위험 없음

### 🔄 데이터베이스 변경

- [ ] 데이터베이스 스키마 변경 없음
- [x] Migration 파일 추가됨
- [x] 기존 데이터와 호환됨 (신규 테이블)
- [x] 롤백 계획 수립됨 (Flyway 버전 관리)

**Migration 상세:**
- `V1__Create_order_tables.sql`
- orders, order_items 테이블 생성
- 7개 인덱스 추가
- CASCADE DELETE 설정

### 📱 호환성

- [x] 이전 버전과 호환됨 (신규 기능)
- [x] Breaking Change 없음
- [x] API 변경 없음
- [ ] 환경 설정 변경 필요

### 📖 문서 업데이트

- [ ] README.md 업데이트 완료
- [ ] API 문서 업데이트 완료
- [x] 코드 주석 업데이트 완료
- [x] 별도 문서 업데이트 불필요 (Repository 계층)

### 🎯 배포 고려사항

- [x] 운영 환경 배포 가능
- [ ] 설정 파일 변경 필요
- [x] 서버 재시작 필요 (신규 배포)
- [x] 데이터 마이그레이션 필요 (Flyway 자동 실행)
- [ ] 특별한 배포 순서 필요

**배포 시 주의사항:**
- Flyway가 자동으로 V1 마이그레이션 실행
- MySQL 9.4 호환성 확인 필요
- JSON 컬럼 타입 지원 확인

### 📸 스크린샷 (UI 변경 시)

해당 없음 (백엔드 Repository 계층)

### 🔍 리뷰 포인트

- **OrderRepositoryImpl.save()** 로직 검토
    - 신규/기존 주문 구분 로직
    - OrderItems cascade 처리 방식
    - 도메인 재구성 로직

- **MapStruct 매핑 전략**
    - 순환 참조 방지 확인
    - 컬렉션 매핑 전략 (toOrderItemDomainList)

- **JPA 쿼리 최적화**
    - JOIN FETCH 쿼리 확인 (findByIdWithItems, findByOrderIdWithItems)
    - 인덱스 활용 여부

- **테스트 커버리지**
    - Mock 기반 단위 테스트 적절성
    - 실패 시나리오 커버리지

### ✅ 체크리스트

- [x] 코드가 컨벤션을 따르고 있습니다
- [x] 자가 리뷰를 완료했습니다
- [x] 적절한 라벨을 추가했습니다
- [x] 관련 문서를 업데이트했습니다
- [x] 테스트를 추가/업데이트했습니다
- [x] CI/CD 파이프라인이 통과합니다

### 💬 추가 메모

**다음 단계:**
- Order Service Application Layer 구현
- Order REST API Controller 구현
- Kafka 이벤트 발행 로직 추가

joel-you committed:
refactor(#57): 57 이슈 리뷰에 대한 반영

* Mapstruct 만 사용하도록 수정
* UUID to Long 로직 제거 (사용하지 않는 로직)

joel-you committed:
feat(#58): Order 데이터베이스 스키마 및 JPA 리포지토리 구현

- Flyway 마이그레이션 스크립트 작성 (V1__Create_order_tables.sql)
- orders, order_items 테이블 생성 with UUID 컬럼
- JPA 엔티티 컬럼명 업데이트 (orderUuid→orderId, userId→customerId)
- 커스텀 쿼리 메서드 구현 (findByCustomerId, findByStatus 등)
- 성능 최적화 인덱스 추가 (customer_id, status, order_date 등)

joel-you committed:
feat(#58): OrderRepositoryImpl save() 메서드 구현

- 신규/기존 주문 저장 및 업데이트 지원
- OrderItems cascade 처리 포함
- MapStruct 기반 도메인 변환 적용

joel-you committed:
feat(#58): OrderRepositoryImpl 단건 조회 메서드 구현

- ID, UUID, 주문번호 기반 조회 지원
- MapStruct 기반 도메인 변환 적용

joel-you committed:
feat(#58): OrderRepositoryImpl 목록 조회 메서드 구현

- 고객/상태/기간 기반 목록 조회 지원
- Stream API 기반 도메인 변환 적용

joel-you committed:
feat(#58): OrderRepositoryImpl Items 함께 조회 메서드 구현

- ID/UUID 기반 주문+항목 조회 지원
- JOIN FETCH로 성능 최적화

joel-you committed:
refactor(#58): Warring 코드 수정

joel-you committed:
refactor(#58): 마이그레이션 파일을 infra/db/migration으로 이동

joel-you committed:
refactor(#58): 마이그레이션 파일을 infra/db/migration으로 이동

joel-you commented on infra/db/migration/V1__Create_order_tables.sql:
```diff
@@ -0,0 +1,61 @@
+    ADD INDEX idx_order_id (order_id),
+    ADD INDEX idx_order_number (order_number),
+    ADD INDEX idx_customer_id (customer_id),
+    ADD INDEX idx_status (status),
```
데이터 분포도 고려해서, 인덱스 추가하는게 좋을 듯

flab-true replied:
데이터를 많이 넣어 보시고, 반드시 실행 계획을 확인해 주세요.

joel-you replied:
@flab-true 님
order_id 의 경우 id와 별개로 따로 uuid 를 잡았던 것 같네요.
따로 리팩터링을 해서, Long값 id를 제거하고, order_id 를 키로 다시 잡도록 하겠습니다.

joel-you commented on infra/db/migration/V1__Create_order_tables.sql:
```diff
@@ -0,0 +1,61 @@
+    DROP INDEX idx_orders_date,
+    ADD INDEX idx_order_id (order_id),
+    ADD INDEX idx_order_number (order_number),
+    ADD INDEX idx_customer_id (customer_id),
```
복합인덱스가 있기 때문에 불필요할듯

joel-you replied:
제거하겠습니다.

joel-you commented on order-orchestrator/src/main/java/com/msa/commerce/orchestrator/adapter/out/persistence/OrderJpaRepository.java:
```diff
@@ -31,17 +31,17 @@ public interface OrderJpaRepository extends JpaRepository<OrderJpaEntity, Long>,
 
-    @Query("SELECT o FROM OrderJpaEntity o WHERE o.userId = :userId AND o.orderDate BETWEEN :startDate AND :endDate")
-    List<OrderJpaEntity> findUserOrdersByDateRange(@Param("userId") Long userId, @Param("startDate") LocalDateTime startDate, @Param("endDate") LocalDateTime endDate);
+    @Query("SELECT o FROM OrderJpaEntity o WHERE o.customerId = :customerId AND o.orderDate BETWEEN :startDate AND :endDate")
```
JPQL -> JPA 메소드로 변환하는게 더 좋을 듯

joel-you replied:
이것도 수정하겠습니다.

joel-you commented on order-orchestrator/src/main/java/com/msa/commerce/orchestrator/adapter/out/persistence/OrderJpaRepository.java:
```diff
@@ -31,17 +31,17 @@ public interface OrderJpaRepository extends JpaRepository<OrderJpaEntity, Long>,
-    long countByUserId(Long userId);
+    long countByCustomerId(Long customerId);
 
     @Query("SELECT DISTINCT o FROM OrderJpaEntity o LEFT JOIN FETCH o.orderItems WHERE o.id = :id")
```
Order 를 찾고, OrderItem 을 가져오면 JQLP 을 대체할 수 있을 것 같음

joel-you commented on order-orchestrator/src/main/java/com/msa/commerce/orchestrator/adapter/out/persistence/OrderRepositoryImpl.java:
```diff
@@ -1,47 +1,67 @@
+        });
+
+        OrderJpaEntity savedEntity = orderJpaRepository.save(orderEntity);
+        return orderMapper.toDomain(savedEntity);
```
코멘트 추가 요망

joel-you replied:
insert 와 save 용도로 같이 쓰게되면서 복잡한 로직이 들어갔는데, insert 전용으로 수정하였습니다.

flab-true commented:
헥사고날 아키텍처에서 의존 역전 원칙(DIP)의 의미를 설명해주세요.

flab-true commented:
헥사고날 아키텍처에서 JPA 영속성 컨텍스트를 함께 사용하는 것은 설계상 어려움이 있습니다. 이 부분에서 trade-off 들이 무엇이 있을지 논의하고 싶어요. 이후에 문서로 정리되면 좋겠습니다.

joel-you committed:
refactor(#65): PR 리뷰 반영 - 인덱스 최적화 및 JPA 개선

- 복합 인덱스 우선 설계로 불필요한 인덱스 제거
- JPQL을 Spring Data JPA 메소드로 변환
- JOIN FETCH를 EntityGraph로 변경

joel-you committed:
refactor(#58): UuidGenerator를 common 모듈로 이동 및 라이브러리 변경

- UUID 라이브러리를 Java UUID Generator(FasterXML)로 변경
- buildSrc에 UUID Generator 의존성 추가하여 중앙 집중식 관리
- UuidGenerator를 common 모듈로 이동하여 모든 서비스에서 재사용 가능하게 개선
- Flyway 마이그레이션 스크립트 에러 수정 (FK, AUTO_INCREMENT, 인덱스 제거 로직 보완)

joel-you committed:
refactor(#65): OrderRepositoryImpl save 메서드 단순화

- update/insert 분기 로직 제거, insert 전용으로 변경
- syncOrderItems 차등 동기화 로직 제거

joel-you commented:
수정완료

joel-you committed:
refactor(#65): V1 마이그레이션 SQL 중복 코드 제거

- 반복되는 인덱스 제거 로직을 stored procedure로 리팩토링
- drop_index_if_exists 프로시저 도입으로 코드 재사용
- SonarCloud 중복 코드 경고 해결

joel-you committed:
refactor(#58): 마이그레이션 SQL 코드 품질 개선

- 테이블명 SQL 변수로 상수화
- UPDATE order_items에 WHERE 조건 추가
- SonarQube Critical 이슈 3건 해결

joel-you commented:
> 헥사고날 아키텍처에서 의존 역전 원칙(DIP)의 의미를 설명해주세요.

[헥사고날 아키텍처에서 의존 역전 원칙(DIP)](https://github.com/f-lab-edu/msa-commerce-lab/wiki/%ED%97%A5%EC%82%AC%EA%B3%A0%EB%82%A0-%EC%95%84%ED%82%A4%ED%85%8D%EC%B2%98%EC%97%90%EC%84%9C-%EC%9D%98%EC%A1%B4-%EC%97%AD%EC%A0%84-%EC%9B%90%EC%B9%99%28DIP%29) 에 기술 해두었습니다.

joel-you commented:
> 헥사고날 아키텍처에서 JPA 영속성 컨텍스트를 함께 사용하는 것은 설계상 어려움이 있습니다. 이 부분에서 trade-off 들이 무엇이 있을지 논의하고 싶어요. 이후에 문서로 정리되면 좋겠습니다.

[헥사고날 아키텍처 JPA 의 문제점](https://github.com/f-lab-edu/msa-commerce-lab/wiki/%ED%97%A5%EC%82%AC%EA%B3%A0%EB%82%A0-%EC%95%84%ED%82%A4%ED%85%8D%EC%B2%98---JPA-%EC%9D%98-%EB%AC%B8%EC%A0%9C%EC%A0%90) 에 작성해두었습니다.

joel-you committed:
Remove .github/README.md from feature/58-order-db-schema-and-jpa-repository

flab-true reviewed (approved):
대단히 고생 많으셨습니다. LGTM 👍
